Adding nav2_dynamic_params package

bpwilcox opened the pull request:
This package has two classes intended to replace part of the functionalies of the ROS1 Dynamic Reconfigure: DynamicParamsValidator & DynamicParamsClient. Addresses #177 

DynamicParamsValidator:
- validates parameter against type (using rclcpp::ParameterType)
- validates parameters against specified upper/lower bounds
- allows for providing custom user validation callback 
- provides interface to add parameters with type and optional bounds (integer & double)

DynamicParamsClient
- passes parameter_client (for remote nodes) and list of "dynamic parameters" to be added from remote parameter server
- keeps cached map of dynamic parameter changes
- provides interface function calls to grab latest parameter change or default value

orduno commented on nav2_dynamic_params/include/nav2_dynamic_params/dynamic_params_client.hpp:
```diff
@@ -0,0 +1,110 @@
+    parameters_client_ = client;
+  }
+
+  void addParametersFromServer(std::vector<std::string> param_names)
```
```cpp
void addParametersFromServer(const std::vector<std::string> & param_names)
```

orduno commented on nav2_dynamic_params/include/nav2_dynamic_params/dynamic_params_client.hpp:
```diff
@@ -0,0 +1,110 @@
+  template<class T>
+  bool get_event_param(
+    const rcl_interfaces::msg::ParameterEvent::SharedPtr event,
+    std::string param_name, T & new_value)
```
```cpp
const std::string & param_name, const T & new_value
```

bpwilcox replied:
`new_value` is not constant

orduno commented on nav2_dynamic_params/include/nav2_dynamic_params/dynamic_params_validator.hpp:
```diff
@@ -0,0 +1,71 @@
+
+  ~DynamicParamsValidator() {}
+
+  void add_param(std::string param_name, rclcpp::ParameterType type);
```
looks like some of the parameters in these functions could be passed as const refs.

orduno commented on nav2_dynamic_params/src/example_dynamic_params.cpp:
```diff
@@ -0,0 +1,88 @@
+  node->set_parameters({rclcpp::Parameter("foo", 1.0), rclcpp::Parameter("bar", 2)});
+
+  // Create DynamicParamsClient
+  dynamic_params_client = new nav2_dynamic_params::DynamicParamsClient(parameters_client);
```
forgot ```delete```

orduno commented on nav2_dynamic_params/src/example_dynamic_params.cpp:
```diff
@@ -0,0 +1,88 @@
+  dynamic_params_client->addParametersFromServer({"foo", "bar"});
+
+  // Create DynamicParamsValidator
+  auto param_validator_ = new nav2_dynamic_params::DynamicParamsValidator(node);
```
forgot ```delete```

crdelsey commented on nav2_dynamic_params/include/nav2_dynamic_params/dynamic_params_client.hpp:
```diff
@@ -0,0 +1,110 @@
+  }
+
+  template<class T>
+  bool get_event_param(
```
This is nearly identical to the previous function. Can you compose the one out of the other to eliminate the duplicate code?

crdelsey replied:
This comment still applies. I think these functions are too similar and should have the commonality factored out in some way.

crdelsey commented on nav2_dynamic_params/include/nav2_dynamic_params/dynamic_params_validator.hpp:
```diff
@@ -0,0 +1,71 @@
+public:
+  explicit DynamicParamsValidator(rclcpp::Node::SharedPtr node);
+
+  ~DynamicParamsValidator() {}
```
I'd leave this line out and let the compiler insert the default destructor.

crdelsey commented on nav2_dynamic_params/src/dynamic_params_validator.cpp:
```diff
@@ -0,0 +1,158 @@
+      "Parameter Change Request: %s", (parameter.get_name()).c_str());
+
+    // Validate Parameter Type
+    for (std::map<std::string, rclcpp::ParameterType>::iterator it = param_map_.begin();
```
This duplicate with the block below. Should be a function. Even if you can't factor out the commonality easily, these blocks make this function too long and should be two functions.

crdelsey replied:
Why are you iterating over the entire map for each parameter? That defeats the purpose of a map. Just look up the key by name. If it's not present, fail it, otherwise call the two validate functions.

crdelsey commented on nav2_dynamic_params/src/dynamic_params_validator.cpp:
```diff
@@ -0,0 +1,158 @@
+  auto result = rcl_interfaces::msg::SetParametersResult();
+  result.successful = true;
+
+  for (auto parameter : parameters) {
```
```suggestion
  for (auto& parameter : parameters) {
```

crdelsey commented on nav2_dynamic_params/src/dynamic_params_validator.cpp:
```diff
@@ -0,0 +1,158 @@
+    }
+
+    // Validate Parameter Bounds
+    for (std::map<std::string, std::pair<double, double>>::iterator it = param_bound_map_.begin();
```
Use range based for here and above
```suggestion
    for (const auto& i :  param_bound_map_)
```

crdelsey commented on nav2_dynamic_params/src/dynamic_params_validator.cpp:
```diff
@@ -0,0 +1,158 @@
+  std::string param_name, rclcpp::ParameterType type,
+  std::pair<double, double> bounds)
+{
+  param_map_[param_name] = type;
```
Rather than two maps, you should just use one map with a structure containing all the data you need.

crdelsey commented on nav2_dynamic_params/src/dynamic_params_validator.cpp:
```diff
@@ -0,0 +1,158 @@
+  std::string param_name, rclcpp::Parameter param,
+  rclcpp::ParameterType type)
+{
+  if (param.get_name() == param_name) {
```
If you use the map to look up the type by name, then you can skip this param_name check.

crdelsey commented on nav2_dynamic_params/src/dynamic_params_validator.cpp:
```diff
@@ -0,0 +1,158 @@
+  std::pair<double, double> bound)
+{
+  if (param.get_name() == param_name) {
+    if (param.get_type() == rclcpp::ParameterType::PARAMETER_DOUBLE) {
```
This code is duplicate with the block below. Make it a function or something.

crdelsey commented on nav2_dynamic_params/src/dynamic_params_validator.cpp:
```diff
@@ -0,0 +1,158 @@
+{
+  if (param.get_name() == param_name) {
+    if (param.get_type() == rclcpp::ParameterType::PARAMETER_DOUBLE) {
+      if (param.get_value<rclcpp::ParameterType::PARAMETER_DOUBLE>() >= bound.first &&
```
A temporary variable here is appropriate.
eg. Doesn't this look better?
```c++
auto p = param.get_value<rclcpp::ParameterType::PARAMETER_DOUBLE>();
if (p >= bound.first && p <= bound.second)

crdelsey commented on nav2_dynamic_params/src/example_dynamic_params.cpp:
```diff
@@ -0,0 +1,88 @@
+
+nav2_dynamic_params::DynamicParamsClient * dynamic_params_client;
+
+int main(int argc, char ** argv)
```
This could have been an automated unit test instead.

crdelsey replied:
Probably should put this in first, as is.

bpwilcox commented:
My latest commit should address the feedback above. Now, the parameter client(s) are created internally when the node is passed into the constructor for the DynamicParamsClient. Thus, it should support listening to events and caching parameters from any number of nodes. In addition, by default I made it such that all parameters on the remote nodes are initialized in the map (I welcome discussion on if this should not be the case). I did this because in ROS2 technically every parameter is dynamic and thus it can be up to the validator to decide whether parameters shouldn't change and up to the callbacks to handle the parameters it cares about. I thought it was a simpler interface that way than manually adding "dynamic" parameters. That being said, I added a function to add static parameters to the DynamicParamsValidator and a check to reject any static parameters. Also, I added a flag to allow for any new parameters to be automatically rejected.

bpwilcox commented:
Though my example node should demonstrate the basic use cases, I still need to write tests for this package. You can delay approval if you feel it should be added before merging, otherwise I'll get to committing those tests when I can.

mjeronimo commented on nav2_dynamic_params/README.md:
```diff
@@ -0,0 +1,142 @@
+In ROS2, rather than on a parameter server, parameters are hosted on the nodes themselves. Nodes may get and set their own
+parameters via the node interface or may create a parameter client to any remote node hosting parameters. By default, any client to
+a node may get and set parameters to that node. A node, however, may register a callback for parameter change requests and choose
+to accept or deny such requests based on arbitrary criteria. In this vain, nodes are responsible for the validation of their own
```
"vain" -> "vein"

bpwilcox commented on nav2_dynamic_params/include/nav2_dynamic_params/dynamic_params_client.hpp:
```diff
@@ -0,0 +1,160 @@
+
+  ~DynamicParamsClient() {}
+
+  void add_parameter_clients(std::vector<std::string> remote_names = {""})
```
Yes, this is because an empty string is interpreted in the SyncParameterClient constructor as creating a client for its own node.

SteveMacenski commented:
Quick note from the weekly discussion: This looks really amazing. Great work! I really look forward to learning this tool more deeply and utilizing it. 

Correct me if I'm misinterpreting the issue we discussed: One "hacky" way to know what node a parameter came from might be to add the node name as a parameter namespace internally to the dynamic parameters package, since you're giving it a node with get_name() method. When you shooting it back out to clients, then you drop that first namespace you artificially added but can use to figure out who to send the updates to.

bpwilcox commented:
> Quick note from the weekly discussion: This looks really amazing. Great work! I really look forward to learning this tool more deeply and utilizing it.
> 
> Correct me if I'm misinterpreting the issue we discussed: One "hacky" way to know what node a parameter came from might be to add the node name as a parameter namespace internally to the dynamic parameters package, since you're giving it a node with get_name() method. When you shooting it back out to clients, then you drop that first namespace you artificially added but can use to figure out who to send the updates to.

Thanks! I appreciate your feedback as well! 

I'm trying to understand your suggestion. I agree that internally we could map the parameter names under the node namespace. However, I'm unsure how you mean we could resolve the source node of a parameter event during the callback. From what I've gathered, in the event, you receive a list of new/changed parameters, which you can filter by name, but there's no info in the event as to where it came from. Are you saying that the client would remap (not sure yet how to do dynamic remapping) the parameters on the host node as well? Please clarify if you can.

crdelsey commented on nav2_dynamic_params/src/dynamic_params_validator.cpp:
```diff
@@ -0,0 +1,165 @@
+    }
+
+    // Validate Parameter Type
+    if (param_map_.count(parameter.get_name()) > 0) {
```
Don't check for this. Just let the exception get thrown. This should never happen.

bpwilcox replied:
I think it is reasonable that someone may attempt to set a new parameter, or perhaps mispelled a parameter. This check would allow us to to accept or reject it based on the flag `reject_new_params_`

crdelsey replied:
Ok. I misunderstood the context of this function. Ok as is.

crdelsey commented on nav2_dynamic_params/src/dynamic_params_validator.cpp:
```diff
@@ -0,0 +1,165 @@
+
+    // Validate Parameter Type
+    if (param_map_.count(parameter.get_name()) > 0) {
+      if (validate_param(parameter)) {
```
Instead of anding together the succesfuls, just
```c++
if(!validate_param(parameter)) {
 result.successful = false;
 return result;
}
```

You could just do this everywhere and simplify the code a bit. There's no point in continuing once you've got any failure.

You could use the pattern below which is pretty common
```c++
  auto result = rcl_interfaces::msg::SetParametersResult();
  result.successful = false;

 for ((auto parameter : parameters) {
  if(!check1(parameter)
    return;
  if(!check2(parameter)
   return;
  if(!check3(parameter)
   return
  }
}
if( validation_callback)
   return validation_callback(parameters)
else
 return result;
```

bpwilcox replied:
I agree we could return sooner. I suppose the only value to allowing it to run the full checks is to get more log information as to additional reasons it would fail validation.

crdelsey replied:
I've been informed that there is a precedent set in the ROS 2 code base for this method of dealing with status messages. I think the pattern I described is more common, but I'm OK with this as is.

SteveMacenski commented:
@bpwilcox bare in mind I haven't read the ROS2 parameter event code before.

But imagine as if you were to have a parameter `foo` in node named `slam_node_or_something` via `rclcpp::Node("slam_node_or_something")`. Inside of the dynamic parameter libraries when you register the parameters you're interested in, you could take them, and tack on the namespace `slam_node_or_something` in front of `foo` to make `slam_node_or_something/foo`. This is what you send over the pipe to the other clients/validators. Which then can use the lowest namespace `slam_node_or_something` to know what node it came from, then strip off the namespace to leave just `foo` for whatever `second_node` updates needing `foo`

bpwilcox commented on nav2_dynamic_params/include/nav2_dynamic_params/dynamic_params_client.hpp:
```diff
@@ -0,0 +1,160 @@
+      {
+        new_value = dynamic_param_map_[param_name].get_value<T>();
+      } else {
+        RCLCPP_DEBUG(rclcpp::get_logger("dynamic_params_client"),
```
What I was going for here was that if the parameter has not been set, then it would not change your value (here `new_value`), and return to the user the result bool so they can do whatever they want with it. I'm not sure whether we'd want to throw an exception. If you always wanted to return a valid value, one could use the overloaded version with a defualt argument. I do understand how this could be confusing to the user, though.

crdelsey replied:
OK. Let's leave it as is then. Let's see how it goes as we apply this code to the rest of the codebase, and see how we like the interface.

mkhansen-intel commented:
Adding OSRF team members: @tfoote @dirk-thomas @wjwwood - we would like you to take a look at this. Please read the enclosed README file and comment.

crdelsey commented on nav2_dynamic_params/include/nav2_dynamic_params/dynamic_params_client.hpp:
```diff
@@ -100,53 +100,64 @@ class DynamicParamsClient
+    }
+  }
+
+  bool check_in_event(
```
Sounds like you want to checkin the event to git or something. A more standard name would be something like
`is_in_event()`. Then the if statement reads like
if(is_in_event())

dirk-thomas commented:
As a temporary workaround the current implementation sounds reasonable. I hope the consensus is that this gets replaced as soon as the functionality has been refactored / ported into the client library (`rcl` or the language specific variations of it).
